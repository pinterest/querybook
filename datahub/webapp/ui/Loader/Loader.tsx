import React from 'react';
import { formatError } from 'lib/utils/error';

import { Loading } from 'ui/Loading/Loading';
import { ErrorMessage } from 'ui/Message/ErrorMessage';
import { useLoader } from 'hooks/useLoader';

export interface ILoaderProps {
    // We check the value of item to see if it needs to be loaded, must not be null
    item: any;

    // Something unique identifies the item, can be undefined as well
    itemKey?: any;

    itemLoader?: () => any;
    // Clean up the loading if there is any
    itemUnloader?: () => any;

    // custom
    placeHolder?: React.ReactNode;
    // Custom renderer function that lets user render after load
    renderer?: () => React.ReactNode;
    // Custom renderer when itemLoader throws an exception
    errorRenderer?: (error: any) => React.ReactNode;
    // Custom renderer when null gets loaded
    emptyRenderer?: () => React.ReactNode;

    children?: any;
}

export const Loader: React.FunctionComponent<ILoaderProps> = ({
    item,
    itemKey,
    itemLoader,
    itemUnloader,
    placeHolder,
    renderer,
    errorRenderer,
    emptyRenderer,
    children,
}) => {
    const { loading, hasError, errorObj } = useLoader({
        item,
        itemKey,
        itemLoader,
        itemUnloader,
    });

    if (hasError) {
        return errorRenderer ? (
            errorRenderer(errorObj)
        ) : (
            <ErrorMessage>{formatError(errorObj)}</ErrorMessage>
        );
    } else if (loading) {
        return placeHolder ? placeHolder : <Loading />;
    } else if (item == null) {
        // We already tried to load and nothing gets returned;
        // temporaily just return null
        return emptyRenderer ? emptyRenderer() : null;
    }

    return renderer ? renderer() : children;
};
