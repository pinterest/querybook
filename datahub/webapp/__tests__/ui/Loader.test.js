import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { Loader } from '../../ui/Loader/Loader';

it('renders without crashing', () => {
    shallow(<Loader />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<Loader />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot', () => {
        const wrapper = shallow(<Loader item="test">children</Loader>);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(<Loader />);
        expect(output).toMatchSnapshot();
    });
});
