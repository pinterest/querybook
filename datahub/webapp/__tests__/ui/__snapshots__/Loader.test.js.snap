// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `<Loading />`;

exports[`matches enzyme snapshots matches snapshot 2`] = `"children"`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<div
  className="sc-bwzfXH bEeIVQ"
>
  <div
    className="sc-htpNat dlDikl"
  >
    <i
      className="fa fa-spinner fa-pulse"
    />
  </div>
</div>
`;
